### PR TITLE
security/acme-client: Add domains.google support

### PR DESCRIPTION
**Pull request by loosecannon93**
Adds support for Google Domains DNS-01 ACME challenge API added to latest upstream acme.sh. This is separate from the Google Cloud API.

https://domains.google/learn/gts-acme/

See https://github.com/acmesh-official/acme.sh/pull/4542

This should address #3408 but ultimately depends on acme.sh cutting a new release with the above pull request included, so I assume would not be able to be merged until then.

### CODE REVIEW

File: security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php
Comment (fraenki) [inline security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php:3]: Please add your copyright information here.

File: security/acme-client/src/opnsense/mvc/app/models/OPNsense/AcmeClient/AcmeClient.xml
Comment (fraenki) [inline security/acme-client/src/opnsense/mvc/app/models/OPNsense/AcmeClient/AcmeClient.xml:677]: I don't think we should make the API endpoint configurable. It adds a config option that all users have to change, if Google decides to change the API endpoint URL. We should [rely on acme.sh to provide the correct URL](https://github.com/acmesh-official/acme.sh/blob/0d25f7612bf37a42d9c4fcb1abc493b2a5e495c3/dnsapi/dns_googledomains.sh#L8) instead.

So I suggest to remove the parameter `dns_googledomains_api_endpoint`.

**Review by fraenki — CHANGES_REQUESTED** [review]

**Comment by fraenki** [comment]
> This should address https://github.com/opnsense/plugins/issues/3408 but ultimately depends on acme.sh cutting a new release with the above pull request included, so I assume would not be able to be merged until then.

Exactly. I'll monitor acme.sh release and will merge it as soon as possible.

File: security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php
Comment (loosecannon93) [reply security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php:3]: done

File: security/acme-client/src/opnsense/mvc/app/models/OPNsense/AcmeClient/AcmeClient.xml
Comment (loosecannon93) [reply security/acme-client/src/opnsense/mvc/app/models/OPNsense/AcmeClient/AcmeClient.xml]: ACK, that makes sense.

I was adding it for completeness, but ideally it would be a hidden or advanced option but I don't know what's possible here. 

Made a note in a comment in the php file that it is technically configurable should it ever turn out to be useful to add it.

**Review by fraenki — APPROVED** [review]

**Comment by dmurphynj** [comment]
Hi team, found an issue with this commit.

I applied it using "opnsense-patch -c plugins e10a960"
and continued to have errors when issuing a cert:
   "AcmeClient: challenge type not supported: dns_googledomains (0ae7b9bb-f72c-47cd-9d78-3bc51b86d235)"

Long story short, it appears that "security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php"
should be renamed to:
"security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogledomains.php"

There is a case-sensitivity issue in LeValidationFactory.php, lines 69-70:
 ```
 // Convert to PascalCase
 $val_name = str_replace(' ', '', ucwords(str_replace(array('-', '_'), ' ', $search_name)));
```
returns as "DnsGoogledomains" after parsing.

Simply renaming DnsGoogleDomains.php to DnsGoogledomains.php seems to work just fine.

os-acme-client plugin 3.17
OPNsense 23.1.9-amd64
FreeBSD 13.1-RELEASE-p7

Hope this helps!  Renaming the php logic solved my issue - works absolutely great now.

**Comment by dmurphynj** [comment]
> There is a case-sensitivity issue in LeValidationFactory.php, lines 69-70:
> 
> ```
> // Convert to PascalCase
> $val_name = str_replace(' ', '', ucwords(str_replace(array('-', '_'), ' ', $search_name)));
> ```
> 
> returns as "DnsGoogledomains" after parsing.

Just to show my I'm-not-a-programmer-and-don't-play-one-on-TV debugging ...  I threw a couple error logging lines around the PascalCase code:

```
    // Convert to PascalCase
        LeUtils::log_error("**dmurphynj: search_name is $search_name");
        $val_name = str_replace(' ', '', ucwords(str_replace(array('-', '_'), ' ', $search_name)));
        LeUtils::log_error("**dmurphynj: val_name is $val_name");
```

From the ACME Client logfile:

```
2023-06-15T18:29:25-04:00 | opnsense | AcmeClient: challenge type not supported: dns_googledomains (9f80a3c6-6892-4a91-b3ac-e867df34ad3c)
-- | -- | --
2023-06-15T18:29:25-04:00 | opnsense | AcmeClient: **dmurphy: val_name is DnsGoogledomains
2023-06-15T18:29:25-04:00 | opnsense | AcmeClient: **dmurphy: search_name is dns_googledomains
2023-06-15T18:29:25-04:00 | opnsense | AcmeClient: account is registered: Google Domains
```

**Comment by xArthasx** [comment]
Given that all other files follow the same convention of being "DnsX$(lowercaseRest)", @loosecannon93 can you please update the PR? cc @fraenki @nmiller0113

File: security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php
Comment (fraenki) [inline security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php:1]: As already pointed out by others, this file needs to be renamed to `DnsGoogledomains.php`.

**Review by fraenki — CHANGES_REQUESTED** [review]

**Comment by fraenki** [comment]
I'll merge this when the requested change was implemented. acme.sh version 3.0.6 added support for Google Domains.

File: security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php
Comment (nmiller0113) [reply security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeValidation/DnsGoogleDomains.php:1]: @fraenki who would do that? I'm happy to help in whatever way I can :) Thanks!

**Comment by nmiller0113** [comment]
@fraenki is there anything I can do to help move this along? I feel like we've been stuck. I'm happy to help, just not sure what to do. Thanks!

**Comment by nmiller0113** [comment]
Or are you waiting on @loosecannon93, @fraenki?

**Comment by nmiller0113** [comment]
> I think that the issue is that github doesn't allow taking over a PR (AFAIK) so either we have to recreate the entire diff or wait for the owner. At this point I'm hesitant on waiting for the creator to do it, wdyt we should do @fraenki ?

Not sure @xArthasx, only that this has sat dormant for weeks now, and not even @fraenki has commented. Maybe he's just not tracking this right now...who knows? All the pieces are there for this to be completed and for the challenge to be added for opnsense...not really sure what the hold up is.

**Comment by xArthasx** [comment]
Added a new PR for this #3499 as owner is unresponsive cc @nmiller0113

**Comment by nmiller0113** [comment]
> Added a new PR for this #3499 as owner is unresponsive cc @nmiller0113

Awesome, thanks!

**Comment by xArthasx** [comment]
@fraenki can you help us move #3499 forward?

**Pull request closed without merging**
